File: app/src/main/java/com/zwstudio/logicgamesandroid/games/slitherlink/android/SlitherLinkGameActivity.java
package com.zwstudio.logicgamesandroid.games.slitherlink.android;

import com.zwstudio.logicgamesandroid.R;
import com.zwstudio.logicgamesandroid.common.android.GameActivity;
import com.zwstudio.logicgamesandroid.common.domain.Position;
import com.zwstudio.logicgamesandroid.games.slitherlink.data.SlitherLinkDocument;
import com.zwstudio.logicgamesandroid.games.slitherlink.data.SlitherLinkMoveProgress;
import com.zwstudio.logicgamesandroid.games.slitherlink.domain.SlitherLinkGame;
import com.zwstudio.logicgamesandroid.games.slitherlink.domain.SlitherLinkGameMove;
import com.zwstudio.logicgamesandroid.games.slitherlink.domain.SlitherLinkGameState;
import com.zwstudio.logicgamesandroid.games.slitherlink.domain.SlitherLinkObject;
import com.zwstudio.logicgamesandroid.games.slitherlink.domain.SlitherLinkObjectOrientation;

import org.androidannotations.annotations.EActivity;

import java.util.List;

@EActivity(R.layout.activity_slitherlink_game)
public class SlitherLinkGameActivity extends GameActivity<SlitherLinkGame, SlitherLinkDocument, SlitherLinkGameMove, SlitherLinkGameState> {
    public SlitherLinkDocument doc() {return app.slitherlinkDocument;}

    protected void startGame() {
        String selectedLevelID = doc().selectedLevelID;
        List<String> layout = doc().levels.get(selectedLevelID);
        tvLevel.setText(selectedLevelID);

        levelInitilizing = true;
        game = new SlitherLinkGame(layout, this);
        try {
            // restore game state
            for (SlitherLinkMoveProgress rec : doc().moveProgress()) {
                SlitherLinkGameMove move = new SlitherLinkGameMove();
                move.p = new Position(rec.row, rec.col);
                move.objOrientation = SlitherLinkObjectOrientation.values()[rec.objOrientation];
                move.obj = SlitherLinkObject.values()[rec.obj];
                game.setObject(move);
            }
            int moveIndex = doc().levelProgress().moveIndex;
            if (!(moveIndex >= 0 && moveIndex < game.moveCount())) return;
            while (moveIndex != game.moveIndex())
                game.undo();
        } finally {
            levelInitilizing = false;
        }
    }
}

File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/slitherlink/SlitherLinkGameActivity.kt
package com.zwstudio.logicpuzzlesandroid.puzzles.slitherlink

import com.zwstudio.logicpuzzlesandroid.R
import com.zwstudio.logicpuzzlesandroid.common.android.GameGameActivity
import com.zwstudio.logicpuzzlesandroid.common.data.GameLevel
import org.androidannotations.annotations.AfterViews
import org.androidannotations.annotations.Bean
import org.androidannotations.annotations.Click
import org.androidannotations.annotations.EActivity

@EActivity(R.layout.activity_game_game)
class SlitherLinkGameActivity : GameGameActivity<SlitherLinkGame, SlitherLinkDocument, SlitherLinkGameMove, SlitherLinkGameState>() {
    @Bean
    protected lateinit var document: SlitherLinkDocument
    override val doc get() = document

    @AfterViews
    protected override fun init() {
        gameView = SlitherLinkGameView(this)
        super.init()
    }

    override fun newGame(level: GameLevel) =
        SlitherLinkGame(level.layout, this, doc)

    @Click
    protected fun btnHelp() {
        SlitherLinkHelpActivity_.intent(this).start()
    }
}